### ISSUE TITLE
NCHW format for Backends to work with Glow 

### ISSUE THREAD

**Issue by dinkdeep** [pytorch/glow#4575]
Looks like Glow seems to work with NHWC formats rather then NCHW formats for which it seems to add a number of Transposes and Reshapes to convert from NCHW->NHWC format which is evident from the looking at the  torch_glow code and loader code where the transposes are added before the nodes o get these conversions. 
Some Backends might prefer to work with NCHW  format and they don't necessarily need these Transposes. 

Eventually the Backends need to write a Pass to spot a Transpose, recreate/replicate a new Node and  use replaceAllUsesOfWith() and get rid of the transpose, This ends up as an exercise where all the transposes needs to be unrolled back .

Is there a way to get all the Nodes in NCHW format , Or Can there be a callback function(similar to shouldLower() ) which is exposed to the Backend , where Glow can call the Backend while running the optimisation passes  for ex when trying to sink a transpose from Input side  to Result side there is an opportunity , where Glow can ask the Backend to handle the transpose, or may be  can remove the transpose depending on user suggestion .

**Comment by jfix71** [comment]
I'm not totally sure the effective ask here -- are you suggesting that we would have an optimization that, when sinking transposes, if it finds an NHWC Convolution with transposed inputs, it would query the backend to find out if it supports NCHW, and if so would simply merge the transposes into the Convolution and change it to an NCHW Convolution?

Just to make sure we're on the same page of current Glow functionality, we have helpers such as e.g. `convertConvToNCHWConv()` which does this conversion for you by adding in transposes to inputs/outputs and changing the layout of the Convolution to NCHW. And then optimizations sink/eliminate as many transposes as possible (should be nearly all of them, depending on the model, and if not all them we probably would just need to add more sinking support).

**Comment by dinkdeep** [comment]
Hi Jordan Thanks for the pointing it out .
"I'm not totally sure the effective ask here" , Indeed it was the ask here as you pointed, essentially I need to run the entire model in NCHW format . 

Looks like these utility functions will come more handy as it seems to support additional Grad Nodes like convertConvGradToNCHWConvGrad,  convertAvgPoolGradToNCHWPool, convertMaxPoolGradToNCHWPool. 

As you pointed out I will try to run on a differentiated resnet18 model coming out of torch_glow and see if it is able to remove all the transposes(Have been using the similar  approach, to get rid of transposes but noticed that some how it was not able to get rid of all the transposes and had to selectively identify the nodes and then add custom conditional's ) Any how I will retry using this utility functions and  check out once again .

Did a grep and noticed that OpenCL seems to be using it, BTW Can I apply the same transformPostLowering() used in OpenCL Backend to the Interpreter Backend, Will it be able to work.The reason being we use NCHW format and Interpreter as a Golden reference most of the times which seems to run in NHWC and the comparison does not comes out one-one and is bit clumsy. So the intent would to get an "NCHW Interepreter", 

On an another Note, I haven't tried OpenCL backend, Does it needs any additional HW or intel specific board or a card which needs to be brought and added, Does it runs out of box like Interpreter backend with just SW installations like opencl libraries only. 

Apologies may  sound a bit naive and dumb here, What does Glow uses as Golden reference to compare the outputs of CPU and Interpreter Backends , would it it pytorch ?

(base) ddeepani@hb-dt-d-deepani:/data/Work/newglow/nife/glow/lib$ grep -r "convertConvToNCHWConv" *
Backends/OpenCL/Transforms.cpp:      auto *NR = convertConvToNCHWConv(CN, F);

**Comment by jfix71** [comment]
>BTW Can I apply the same transformPostLowering() used in OpenCL Backend to the Interpreter Backend, Will it be able to work.The reason being we use NCHW format and Interpreter as a Golden reference most of the times. Interpreter runs in NHWC and the comparison outputs do not come out one-one and are bit clumsy. So the intent would to get an "NCHW Interepreter" for any easy inter-operable reference.

No, this won't work, as the Interpreter backend only supports NHWC. I'm not sure why you have an issue with this though -- for example we have many OperatorTests which use the helper `compareAgainstInterpreter()`, which does a comparison of results of e.g. OpenCL against Interpreter. Any modification of the layout shouldn't make a numerical difference, and if the output Placeholder is NHWC then the NCHW result will be transposed back to NHWC before being saved to the output Placeholder.

>On an another Note, I haven't tried OpenCL backend, Does it needs any additional HW or intel specific board or a card which needs to be brought and added, Does it runs out of box like Interpreter backend with just SW installations like opencl libraries only.

It should run fine even on your CPU, as long as you have OpenCL libraries installed. You can actually switch between different OpenCL devices available using `-device` (see [here](https://github.com/pytorch/glow/blob/d4fce38db2ba0c4a52f8ce1de02003b722099fa6/lib/Backends/OpenCL/OpenCL.cpp#L90-L94)).

>Apologies may sound a bit naive and dumb here, What does Glow uses as Golden reference to compare the outputs of CPU and Interpreter Backends , would it it pytorch ?

The Interpreter is generally used as the reference/golden result. We could use PyTorch as well for some use cases, though note PyTorch support was added relatively recently compared to original Glow. We also have some support for comparing numerics against Caffe2. However in general there's no perfect way to do this, as different backends/frameworks can choose to implement kernels differently in valid ways, and so there will always inevitably be some numerical differences.

**Comment by dinkdeep** [comment]
Thanks Jordan for pointing to the function compareAgainstInterpreter() this indeed helps , As you pointed it looks like this compares the place holders only , Can it be extended to cater to compare the tensors resulting from intermediate allocactivations, Some times in bigger network with large number of Nodes we would like to debug the inputs & outputs of inner nodes,i.e there  intermediate Tensors being allocated from allocactivations as interpreter does using
void BoundInterpreterFunction::fwdAllocActivationInst( const AllocActivationInst *I) {getOrCreateTensor(I);}

Sometimes these kind of conditions may arise despite the fact that every single node has been verified but on integrating these nodes, the final output does not matches and in case we like to traverse each node and ensure that its input/outputs are all crisp  (and am referring to the problem in our custom backend driver, not the interpreter driver , meaning a bug in our custom driver which wrongly programs the state in a integrated environment at some inner node, the problem manifests at an output place holder, but the root problem was at some intermediate Node, If I have a snapshots of all the tensors [including Placeholder/Constants/Alloca ]then this might help a bit to spot the offending Node and its associated Tensor).

"The Interpreter is generally used as the reference/golden result", Just wanted to understand the methodologies you guys follow 
Assuming that you don't want to refer to any of external references for Golden, Do you manually generate the  inputs/outputs feeds using pencil/paper,  Take for ex this equation which is one output of NormalizationGradNode inside the lowerBatchNormalizationGradNode()
dh = (1. / N) * gamma * (var + eps)**(-1. / 2.) * (N * dy - np.sum(dy) - (h - mu) * 1/(var + eps) * np.sum(dy * (h - mu)))
I mean to say we can paper out couple of cases but in some corner cases it might get totally south . 

My point is for a given set of 32 bit float inputs any implementation like pytorch/tensorflow/cntk for the above equation should have identical output with (minor bit differences in fractional values for ex pyt may generate dh = 123.12345 vs tesnorflow can get dh = 123.12300 ) which is acceptable but cntk generating dh =  119.999 needs attention and a revisit to implementation) .
